cpp/src/fil/common.cuh

```diff
@@ -114,6 +114,16 @@ struct sparse_storage : storage_base {
 typedef sparse_storage<sparse_node16> sparse_storage16;
 typedef sparse_storage<sparse_node8> sparse_storage8;
 
+template <typename fil_node_t>
+struct node2storage {
```

Comment:
There are several types like this in this pull request. It would be better to unify them into a single `node_traits` type. E.g.

```
// for sparse nodes
template <typename node_t>
struct node_traits {
  using storage = sparse_storage<node_t>;
  using forest = sparse_forest<node_t>;
  static const bool IS_DENSE = false;
};

// for dense nodes
template<> 
struct node_traits<dense_node> {
  using forest = dense_forest;
  using storage = dense_storage;
  static const bool IS_DENSE = true;
};
```

It is also possible to include the `check()` method that checks whether the number of nodes fits into `sparse_node8`, but that's up to you.

Reply:
Though possible, it won't be as neat, as it would mix the node data structure with unrelated information.

A separate `node_traits` data type looks better.